src/webgpu/api/operation/rendering/blending.spec.ts

```diff
@@ -7,7 +8,269 @@ TODO:
 import { GPUTest } from '../../../gpu_test.js';
 
 export const g = makeTestGroup(GPUTest);
+
+const kBlendFactors: GPUBlendFactor[] = [
+  'zero',
+  'one',
+  'src-color',
+  'one-minus-src-color',
+  'src-alpha',
+  'one-minus-src-alpha',
+  'dst-color',
+  'one-minus-dst-color',
+  'dst-alpha',
+  'one-minus-dst-alpha',
+  'src-alpha-saturated',
+  'blend-color',
+  'one-minus-blend-color',
+];
+
+const kBlendOperations: GPUBlendOperation[] = [
+  'add', //
+  'subtract',
+  'reverse-subtract',
+  'min',
+  'max',
+];
+
+function mapColor(
+  col: GPUColorDict,
+  f: (v: number, k: keyof GPUColorDict) => number
+): GPUColorDict {
+  return {
+    r: f(col.r, 'r'),
+    g: f(col.g, 'g'),
+    b: f(col.b, 'b'),
+    a: f(col.a, 'a'),
+  };
+}
+
+function computeBlendFactor(
+  src: GPUColorDict,
+  dst: GPUColorDict,
+  col: GPUColorDict,
+  factor: GPUBlendFactor
+): GPUColorDict {
+  switch (factor) {
+    case 'zero':
+      return { r: 0, g: 0, b: 0, a: 0 };
+    case 'one':
+      return { r: 1, g: 1, b: 1, a: 1 };
+    case 'src-color':
+      return { ...src };
+    case 'one-minus-src-color':
+      return mapColor(src, v => 1 - v);
+    case 'src-alpha':
+      return mapColor(src, () => src.a);
+    case 'one-minus-src-alpha':
+      return mapColor(src, () => 1 - src.a);
+    case 'dst-color':
+      return { ...dst };
+    case 'one-minus-dst-color':
+      return mapColor(dst, v => 1 - v);
+    case 'dst-alpha':
+      return mapColor(dst, () => dst.a);
+    case 'one-minus-dst-alpha':
+      return mapColor(dst, () => 1 - dst.a);
+    case 'src-alpha-saturated': {
+      const f = Math.min(src.a, 1 - src.a);
+      return { r: f, g: f, b: f, a: 1 };
+    }
+    case 'blend-color':
+      return { ...col };
+    case 'one-minus-blend-color':
+      return mapColor(col, v => 1 - v);
+  }
+}
+
+function computeBlendOperation(src: GPUColorDict, dst: GPUColorDict, operation: GPUBlendOperation) {
+  switch (operation) {
+    case 'add':
+      return mapColor(src, (v, k) => v + dst[k]);
+    case 'max':
+      return mapColor(src, (v, k) => Math.max(v, dst[k]));
+    case 'min':
+      return mapColor(src, (v, k) => Math.min(v, dst[k]));
+    case 'reverse-subtract':
+      return mapColor(src, (v, k) => dst[k] - v);
+    case 'subtract':
+      return mapColor(src, (v, k) => v - dst[k]);
+  }
+}
+
+g.test('GPUBlendComponent')
+  .desc(
+    `Test all combinations of parameters for GPUBlendComponent.
+
+  Tests that parameters are correctly passed to the backend API and blend computations
+  are done correctly by blending a single pixel. The test uses rgba32float as the format
+  to avoid checking clamping behavior (tested in api,operation,rendering,blending:clamp,*).
+
+  Params:
+    - component= {color, alpha} - whether to test blending the color or the alpha component.
+    - srcFactor= {...all GPUBlendFactors}
+    - dstFactor= {...all GPUBlendFactors}
+    - operation= {...all GPUBlendOperations}
+  `)
+  .cases(
```

Comment:
Not sure where this should be, but we should leave a TODO _somewhere_ for blending into srgb render targets. (Just thought of this while [looking at](https://www.khronos.org/registry/vulkan/specs/1.2/html/chap27.html#framebuffer-blendoperations) the Vulkan spec.)

Reply:
Oh, and something to make sure blending uses src0 properly (in case an implementation reorders render targets or something).